syntax = "proto3";
import "google/protobuf/empty.proto";
import "google/protobuf/any.proto";
import "sharedMessages.proto";
package dk.itu.real.ooe;

/**
    The entity service
*/
service EntityService {
    /** Reads multiple entities **/
    rpc readEntities (Uuids) returns (Entities);

    /** Spawn multiple entities. */
    rpc spawnEntities (SpawnEntities) returns (Uuids);

    /** Read entities in specified radius */
    rpc readEntitiesInSphere(Sphere) returns (Entities);

    /** Set the AI Tasks for an entity */
    rpc updateEntityAI(EntityAIUpdate) returns (google.protobuf.Empty);
}

/** Unique ids of spawned entities */
message Uuids {
    repeated string uuids = 1;
}

/** A collection of entities */
message Entities {
    repeated Entity entities = 1;
}

/** A single entity */
message Entity {
    string id = 1;
    EntityType type = 2;
    Point position = 3;
    bool isLoaded = 4;
}

/** A collection of spawn messages */
message SpawnEntities {
    repeated SpawnEntity spawnEntities = 1;
}

/** Message for spawning entities */
message SpawnEntity {
    EntityType type = 1;
    Point spawnPosition = 2;
}

/** A sphere area in 3d space */
message Sphere {
    Point center = 1;
    double radius = 2;
}

/** A collection of tasks applied to a specific entity */
message EntityAIUpdate {
    string uuid = 1;
    bool resetGoals = 2;
    repeated AITask AITasks = 3;
}

/** A single task and its priority */
message AITask {
    int32 priority = 1;
    google.protobuf.Any task = 2;
}

/** A task that makes entities idle */
message AITask_Idle {}

/** A task that makes an entity watch other entities */
message AITask_WatchClosest {
    float chance = 1;
    float maxDistance = 2;
    EntityType entityType = 3;
}

/** A task that makes an entity attack their targets */
message AITask_AttackLiving {
    double speed = 1;
    bool hasLongMemory = 2;
}

/** A task that makes an entity run away from other types */
message AITask_AvoidEntityTypes {
    double closeRangeSpeed = 1;
    double farRangeSpeed = 2;
    float searchDistance = 3;
    repeated EntityType entityType = 4;
}

/** A task that makes an entity run away other entities based on unique ids */
message AITask_AvoidSpecificEntities {
    double closeRangeSpeed = 1;
    double farRangeSpeed = 2;
    float searchDistance = 3;
    repeated string uuids = 4;
}

/** A task that makes an entity attack with ranged weapons */
message AITask_RangeAgent {
    float attackRadius = 1;
    int32 delayBetweenAttacks = 2;
    double moveSpeed = 3;
}

/** A task that makes an entity wander aimlessly */
message AITask_Wander {
    int32 chance = 1;
    double speed = 2;
}

/** A target selector for entity types */
message AITask_FindNearestTarget {
    int32 chance = 1;
    EntityType targetEntity = 2;
    bool onlyNearby = 3;
    bool shouldCheckSight = 4;
}

/** A target selector based in entity ids */
message AITask_FindSpecificTarget {
    int32 chance = 1;
    repeated string uuids = 2;
    bool onlyNearby = 3;
    bool shouldCheckSight = 4;
}

/** Types of entities that can be spawned in Minecraft 1.12 */
enum EntityType {
    ENTITY_AREA_EFFECT_CLOUD=0;
    ENTITY_ARMOR_STAND=1;
    ENTITY_BAT=2;
    ENTITY_BLAZE=3;
    ENTITY_BOAT=4;
    ENTITY_CAVE_SPIDER=5;
    ENTITY_CHESTED_MINECART=6;
    ENTITY_CHICKEN=7;
    ENTITY_COMMANDBLOCK_MINECART=8;
    ENTITY_COMPLEX_PART=9;
    ENTITY_COW=10;
    ENTITY_CREEPER=11;
    ENTITY_DONKEY=12;
    ENTITY_DRAGON_FIREBALL=13;
    ENTITY_EGG=14;
    ENTITY_ELDER_GUARDIAN=15;
    ENTITY_ENDER_CRYSTAL=16;
    ENTITY_ENDER_DRAGON=17;
    ENTITY_ENDER_PEARL=18;
    ENTITY_ENDERMAN=19;
    ENTITY_ENDERMITE=20;
    ENTITY_EVOCATION_FANGS=21;
    ENTITY_EVOCATION_ILLAGER=22;
    ENTITY_EXPERIENCE_ORB=23;
    ENTITY_EYE_OF_ENDER=24;
    ENTITY_FALLING_BLOCK=25;
    ENTITY_FIREBALL=26;
    ENTITY_FIREWORK=27;
    ENTITY_FISHING_HOOK=28;
    ENTITY_FURNACE_MINECART=29;
    ENTITY_GHAST=30;
    ENTITY_GIANT=31;
    ENTITY_GUARDIAN=32;
    ENTITY_HOPPER_MINECART=33;
    ENTITY_HORSE=34;
    ENTITY_HUMAN=35;
    ENTITY_HUSK=36;
    ENTITY_ILLUSION_ILLAGER=37;
    ENTITY_IRON_GOLEM=38;
    ENTITY_ITEM=39;
    ENTITY_ITEM_FRAME=40;
    ENTITY_LEASH_HITCH=41;
    ENTITY_LIGHTNING=42;
    ENTITY_LLAMA=43;
    ENTITY_LLAMA_SPIT=44;
    ENTITY_MAGMA_CUBE=45;
    ENTITY_MOB_SPAWNER_MINECART=46;
    ENTITY_MULE=47;
    ENTITY_MUSHROOM_COW=48;
    ENTITY_OCELOT=49;
    ENTITY_PAINTING=50;
    ENTITY_PARROT=51;
    ENTITY_PIG=52;
    ENTITY_PIG_ZOMBIE=53;
    ENTITY_PLAYER=54;
    ENTITY_POLAR_BEAR=55;
    ENTITY_PRIMED_TNT=56;
    ENTITY_RABBIT=57;
    ENTITY_RIDEABLE_MINECART=58;
    ENTITY_SHEEP=59;
    ENTITY_SHULKER=60;
    ENTITY_SHULKER_BULLET=61;
    ENTITY_SILVERFISH=62;
    ENTITY_SKELETON=63;
    ENTITY_SKELETON_HORSE=64;
    ENTITY_SLIME=65;
    ENTITY_SMALL_FIREBALL=66;
    ENTITY_SNOWBALL=67;
    ENTITY_SNOWMAN=68;
    ENTITY_SPECTRAL_ARROW=69;
    ENTITY_SPIDER=70;
    ENTITY_SPLASH_POTION=71;
    ENTITY_SQUID=72;
    ENTITY_STRAY=73;
    ENTITY_THROWN_EXP_BOTTLE=74;
    ENTITY_TIPPED_ARROW=75;
    ENTITY_TNT_MINECART=76;
    ENTITY_UNKNOWN=77;
    ENTITY_VEX=78;
    ENTITY_VILLAGER=79;
    ENTITY_VINDICATION_ILLAGER=80;
    ENTITY_WEATHER=81;
    ENTITY_WITCH=82;
    ENTITY_WITHER=83;
    ENTITY_WITHER_SKELETON=84;
    ENTITY_WITHER_SKULL=85;
    ENTITY_WOLF=86;
    ENTITY_ZOMBIE=87;
    ENTITY_ZOMBIE_HORSE=88;
    ENTITY_ZOMBIE_VILLAGER=89;
}